### PR TITLE
[DSL] Add ensure_bundler! flag

### PR DESCRIPTION
**Pull request by seanreinhardtapps**
closes CocoaPods/CocoaPods#9325

### CODE REVIEW

**Comment by amorde** [comment]
Looks like this also needs a rebase

**Comment by dnkoutso** [comment]
Is it possible to extend this `ensure_bundler!` to take a version?

**Comment by seanreinhardtapps** [comment]
Updated with an optional version check.  I tried it out with a semantic version check, so it could perform an exact or minimum version check, but if it would be simpler as a minimum version check, I can easily update it.

File: spec/podfile/dsl_spec.rb [hunk 1]
```diff
@@ -679,6 +679,90 @@ module Pod
 
     #-------------------------------------------------------------------------#
 
+    describe 'Ensure Bundler' do
```
Comment (dnkoutso) [inline spec/podfile/dsl_spec.rb:682]: nit: for all spec names we use lower case.

File: lib/cocoapods-core/podfile/dsl.rb [hunk 2]
```diff
@@ -71,6 +71,36 @@ def install!(installation_method, options = {})
...
+      #
+      def ensure_bundler!(version = nil)
+        unless current_target_definition.root?
+          raise Informative, 'The Ensure Bundler check can only be set at the root level of the Podfile.'
+        end
+        unless %w(BUNDLE_BIN_PATH BUNDLE_GEMFILE).all? { |key| ENV.key?(key) }
+          CoreUI.warn "CocoaPods was invoked from Global Gemset.\nPlease re-run using: `bundle exec pod #{ARGV.join(' ')}`"
+        end
+        unless version.nil? || ENV['BUNDLER_VERSION'].nil? ||
+            Gem::Dependency.new('', version).match?('', ENV['BUNDLER_VERSION'])
```
Comment (dnkoutso) [inline lib/cocoapods-core/podfile/dsl.rb:99]: can you also treat it as `Requirement` and use `requirement_satisfied_by?` to ensure the version is satisfied?

Comment (dnkoutso) [reply lib/cocoapods-core/podfile/dsl.rb:99]: Like 
```ruby
Requirement.create(version).satisfied_by?(ENV['BUNDLER_VERSION'])
```
?

File: CHANGELOG.md [hunk 3]
```diff
@@ -10,6 +10,10 @@
 
 ##### Enhancements
 
+* Added `ensure_bundler!` flag to the Podfile DSL.  
+  [Sean Reinhardt](https://github.com/seanreinhardtapps)
+  [#7432](https://github.com/CocoaPods/CocoaPods/issues/9325)
```
Comment (dnkoutso) [inline CHANGELOG.md:15]: @seanreinhardtapps #7432 does not match the issue 9325 in this line.

File: spec/podfile/dsl_spec.rb [hunk 4]
```diff
@@ -679,6 +679,90 @@ module Pod
...
+      it 'Has no warning with Bundler environment' do
+        stash_bundler_environment
+        ENV['BUNDLE_GEMFILE'] = 'mock argument'
+        ENV['BUNDLE_BIN_PATH'] = 'another mock argument'
+        ENV['BUNDLER_VERSION'] = '1.17.1'
+        Podfile.new do
+          ensure_bundler!
+        end
+        restore_bundler_environment
+        CoreUI.warnings.should == ''
```
Comment (dnkoutso) [inline spec/podfile/dsl_spec.rb:726]: nit: `should.be.empty`?

File: spec/podfile/dsl_spec.rb [hunk 5]
```diff
@@ -679,6 +679,102 @@ module Pod
...
+        ENV['BUNDLE_BIN_PATH'] = 'another mock argument'
+        ENV['BUNDLER_VERSION'] = '2.0.9'
+        Podfile.new do
+          ensure_bundler! '~> 2.0.0'
+        end
+        restore_bundler_environment
+        CoreUI.warnings.should.be.empty
+      end
+
+      it 'Skips version check with missing Bundle Version (edge case)' do
```
Comment (seanreinhardtapps) [inline spec/podfile/dsl_spec.rb:753]: Please double-check this is how we want this case to play out (Version check silently skipped if `ENV['BUNDLER_VERSION']` is nil).

File: CHANGELOG.md (hunk 3 above)
Comment (seanreinhardtapps) [reply CHANGELOG.md]: :man_facepalming:

File: spec/podfile/dsl_spec.rb (hunk 5 above)
Comment (dnkoutso) [reply spec/podfile/dsl_spec.rb:753]: its fine its an edge case.

File: spec/podfile/dsl_spec.rb [hunk 6]
```diff
@@ -679,6 +679,102 @@ module Pod
...
+        @stored_bundle_version = ENV['BUNDLER_VERSION']
+      end
+
+      def restore_bundler_environment
+        ENV['BUNDLE_BIN_PATH'] = @stored_bundle_bin_path
+        ENV['BUNDLE_GEMFILE'] = @stored_bundle_gemfile
+        ENV['BUNDLER_VERSION'] = @stored_bundle_version
+      end
+
+      it 'Prints warning when Bundler environment not found' do
```
Comment (dnkoutso) [inline spec/podfile/dsl_spec.rb:695]: one more nit: For all those `it` here can you do lower case?

**Comment by dnkoutso** [comment]
Looks good!

File: spec/podfile/dsl_spec.rb (hunk 6 above)
Comment (seanreinhardtapps) [reply spec/podfile/dsl_spec.rb]: ✔️

**Pull request merged**